README.md
# nomnoml-cli

[![Latest version](https://img.shields.io/npm/v/nomnoml-cli)
 ![Dependency status](https://img.shields.io/librariesio/release/npm/nomnoml-cli)
](https://www.npmjs.com/package/nomnoml-cli)
[![codecov](https://codecov.io/gh/prantlf/nomnoml-cli/branch/master/graph/badge.svg)](https://codecov.io/gh/prantlf/nomnoml-cli)
[![Code Climate](https://codeclimate.com/github/prantlf/nomnoml-cli/badges/gpa.svg)](https://codeclimate.com/github/prantlf/nomnoml-cli)

Generates images from [nomnoml](http://www.nomnoml.com/) diagram sources
on the command line and provides a library for a programmatic usage from
[NodeJS] modules

## Getting Started

Make sure that you have [NodeJS] >= 12 installed.

1. Install [pre-requisites](https://github.com/Automattic/node-canvas/wiki/_pages)
   of the [node-canvas](https://github.com/Automattic/node-canvas) module depending
   on your operating system

2. Install the command-line tool and generate a testing image:

```bash
npm install -g nomnoml-cli
echo '[nomnoml]is->[awesome]' | nomnoml > awesome.png
```

Read the documentation about the [nomnoml source format](https://github.com/skanaar/nomnoml#example)

## Command-Line Usage

The `nomnoml` script generates a png, jpg, svg or pdf image from the nomnoml source text.
Both file names and standard input and output are supported as parameters.
If generating the image fails, exit code 1 is returned to the caller.

```text
$ nomnoml --help

  Usage: nomnoml [option]

  Options:

    -h, --help             output usage information
    -V, --version          output the version number
    -i, --input <path>     file with nomnoml source to read from
    -o, --output <path>    file for the image output to write to
    -f, --format <format>  output format (png, jpg, svg, pdf)
    -w, --width <pixels>   width of the canvas to draw on
    -H, --height <pixels>  height of the canvas to draw on

  The default output format is png. The default canvas size is 640x480 pixels.
  If the input file is omitted, the source is read from the standard input.
  If the output file is omitted, the image is written to the standard output.

  Examples:

    $ echo '[nomnoml]is->[awesome]' | nomnoml > awesome.png
    $ nomnoml -i source.nomnoml -f svg -o target.svg
```

## Programmatic Usage

The main module exports a function which generates the image:

```javascript
const generateDiagram = require('nomnoml-cli');
const diagram = await generateDiagram(...);
```

The function returns a [Promise] to wait for a success or a failure.

The function expects a nomnoml source text as a `string` or an `options`
object with the following supported properties:

* `input`: source to read the nomnoml text from; either a [Stream] or a
    `string` with the actual nomnoml text
* `inputFile`: source file path to read the nomnoml source from
* `output`: target to write the image to; either a [Stream] or a `string`
    with the file path (undefined by default)
* `resultType`: type of the object to resolve the promise with if the
    `output` parameter is not provided ('buffer' or 'stream'; the former
    is default)
* `format`: output image format ('png', 'jpg', 'svg' or 'pdf'; the first
    is default)
* `width`: canvas width in pixels (640 by default)
* `height`: canvas height in pixels (480 by default)

Either `input` or `inputFile` has to be provided, otherwise the function
fails.  If `output` is not provided, the function resolves the promise
with a [Buffer] containing the image.  If `output` is provided, the
function returns a promise resolved when the output has been written.

### Code Samples

```javascript
const generateDiagram = require('nomnoml-cli');

// Get Buffer with the image
const buffer = await generateDiagram('[nomnoml]is->[awesome]');

// Convert the standard input to standard output
await generateDiagram({
  input: process.stdin,
  output: process.stdout
});

// Create a PNG file from a nomnoml source file
await generateDiagram({
  inputFile: 'source.nomnoml',
  output: 'target.png'
});
```

## Notes

Contents of other `.nomnoml` files can be imported to the input file using
the [`#import` directive](https://github.com/skanaar/nomnoml#directives).
You can use it for sharing style definitions or parts of diagrams among
multiple diagram sources.

```text
#import: /usr/local/share/nomnoml/style.nomnoml
#import: shared/style.nomnoml
#import: ./sub-diagram.nomnoml
```

If the imported file path is relative and starts with `./` or `../`, it will
be appended to the path of the "parent" file, which the specified file is
being imported to. Otherwise the relative path will be appended to the
current (executing) directory.

File import directives are processed recursively. However, a single file can
be imported only once. If imported once more, scond and other occurrences
will be replaced by an empty string.

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding
style.  Add unit tests for any new or changed functionality. Lint and test
your code using Grunt.

## License

Copyright (c) 2015-2022 Ferdinand Prantl

Licensed under the MIT license.

[Buffer]: https://nodejs.org/api/buffer.html
[NodeJS]: http://nodejs.org/
[Stream]: https://nodejs.org/api/stream.html
[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise


CHANGELOG.md
# [3.1.0](https://github.com/prantlf/nomnoml-cli/compare/v3.0.0...v3.1.0) (2022-01-30)


### Features

* Support JPG and PDF as additional output formats ([5e67b16](https://github.com/prantlf/nomnoml-cli/commit/5e67b16386c603e5b028b87a1a60eac942d0822c))

# [3.0.0](https://github.com/prantlf/nomnoml-cli/compare/v2.0.0...v3.0.0) (2022-01-30)


### Bug Fixes

* Upgrade dependencies ([7ab6b54](https://github.com/prantlf/nomnoml-cli/commit/7ab6b541fef2b740e68d983add4d744f7b506fc7))


### BREAKING CHANGES

* The minimum version of Node.js is 12 instead of 8. The `commander` dependency requires at least that version.
* The command line parameter to set the image height is `-H, --height`. The `h` conflicted with the `-h, --help` option.

## 2019-09-22   v2.0.0

Dropped support of Node.js 6

## 2018-04-27   v1.0.0

Dropped support of Node.js 4

## 2017-05-01   v0.6.1

Enable automatic versioning by semantic-release

## 2017-05-01   v0.6.0

Add support for the #import directive

## 2017-05-01   v0.5.1

Swith the nomnoml dependency from my fork to the upstream

## 2017-04-22   v0.5.0

Return exit code 1 in case of failure

## 2017-04-16   v0.4.7

Update dependencies

## 2017-02-23   v0.4.6

Update dependencies

## 2016-12-19   v0.4.3

Add --format to be able to select between PNG and SVG (thanks, Emanuele Aina)

## 2016-26-08   v0.3.0

Upgrade to Grunt 1.x

## 2016-01-09   v0.2.2

Upgrade development dependencies, fix e-mail, update copyright year

## 2015-10-24   v0.2.2

Upgrade development dependencies, test with the latest prantlf/nomnoml#combined

## 2015-08-06   v0.2.1

Fix code coverage computation

## 2015-08-05   v0.2.0

Add the `resultType` option and tests

## 2015-07-31   v0.1.1

Improve the documentation

## 2015-07-30   v0.1.0

Initial release


test/generate_test.js
'use strict';

const { statSync, unlinkSync, createReadStream, createWriteStream } = require('fs');
const { join } = require('path');
const { NOMNOML_CLI_COVERAGE: coverage } = process.env;
const generateDiagram = require(coverage ? '../coverage/lib/generate' : '../lib/generate');

function readStream (stream) {
  const chunks = [];
  return new Promise((resolve, reject) => {
    stream.on('data', chunk => chunks.push(Buffer.from(chunk)));
    stream.on('error', error => reject(error));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
  });
}

exports.when = {
  'called on the command line': function (test) {
    const name = join(__dirname, 'piracy');
    const output = statSync(name + '.png');
    test.ok(output.isFile() && output.size > 0, 'creates a PNG file');
    unlinkSync(name + '.png');
    test.done();
  },

  'called on the command line with format=svg': function (test) {
    const name = join(__dirname, 'piracy');
    const output = statSync(name + '.svg');
    test.ok(output.isFile() && output.size > 0, 'creates a SVG file');
    unlinkSync(name + '.svg');
    test.done();
  },

  'required': function (test) {
    test.equal('function', typeof generateDiagram, 'returns a generator function');
    test.done();
  },

  'called without arguments': function (test) {
    generateDiagram()
      .then(function () {
        test.ok(false, 'should fail');
        test.done();
      })
      .catch(function (error) {
        test.ok(true, 'fails');
        test.ok(error instanceof Error, 'returns an error');
        test.equal('Missing arguments', error.message, 'describes the failure: "Missing arguments"');
        test.done();
      });
  },

  'called without input': function (test) {
    generateDiagram({})
      .then(function () {
        test.ok(false, 'should fail');
        test.done();
      })
      .catch(function (error) {
        test.ok(true, 'fails');
        test.ok(error instanceof Error, 'returns an error');
        test.equal('Missing input', error.message, 'describes the failure: "Missing input"');
        test.done();
      });
  },

  'called with invalid arguments': function (test) {
    generateDiagram(1)
      .then(function () {
        test.ok(false, 'should fail');
        test.done();
      })
      .catch(function (error) {
        test.ok(true, 'fails');
        test.ok(error instanceof Error, 'returns an error');
        test.equal('Invalid arguments', error.message, 'describes the failure: "Invalid arguments"');
        test.done();
      });
  },

  'called with a string expecting a PNG buffer': function (test) {
    generateDiagram('[nomnoml]is->[awesome]')
      .then(function (result) {
        test.ok(true, 'succeeds');
        test.ok(result instanceof Buffer, 'returns a PNG buffer');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with a string expecting a SVG buffer': function (test) {
    generateDiagram({
      input: '[nomnoml]is->[awesome]',
      format: 'svg'
    })
      .then(function (result) {
        test.ok(true, 'succeeds');
        test.ok(result instanceof Buffer, 'returns a SVG buffer');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with a string expecting a PNG stream': function (test) {
    generateDiagram({
      input: '[nomnoml]is->[awesome]',
      resultType: 'stream'
    })
      .then(async function (result) {
        test.ok(true, 'succeeds');
        const buffer = await readStream(result);
        test.equal('P'.charCodeAt(0), buffer[1], 'returns "P" in a PNG stream');
        test.equal('N'.charCodeAt(0), buffer[2], 'returns "N" in a PNG stream');
        test.equal('G'.charCodeAt(0), buffer[3], 'returns "G" in a PNG stream');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with a string expecting a JPG stream': function (test) {
    generateDiagram({
      input: '[nomnoml]is->[awesome]',
      format: 'jpg',
      resultType: 'stream'
    })
      .then(async function (result) {
        test.ok(true, 'succeeds');
        const buffer = await readStream(result);
        test.equal('J'.charCodeAt(0), buffer[6], 'returns "J" in a JPG stream');
        test.equal('F'.charCodeAt(0), buffer[7], 'returns "F" in a JPG stream');
        test.equal('I'.charCodeAt(0), buffer[8], 'returns "I" in a JPG stream');
        test.equal('F'.charCodeAt(0), buffer[9], 'returns "F" in a JPG stream');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with a string expecting an SVG stream': function (test) {
    generateDiagram({
      input: '[nomnoml]is->[awesome]',
      format: 'svg',
      resultType: 'stream'
    })
      .then(async function (result) {
        test.ok(true, 'succeeds');
        const buffer = await readStream(result);
        test.equal('<'.charCodeAt(0), buffer[0], 'returns "<" in a SVG stream');
        test.equal('?'.charCodeAt(0), buffer[1], 'returns "?" in a SVG stream');
        test.equal('x'.charCodeAt(0), buffer[2], 'returns "x" in a SVG stream');
        test.equal('m'.charCodeAt(0), buffer[3], 'returns "m" in a SVG stream');
        test.equal('l'.charCodeAt(0), buffer[4], 'returns "l" in a SVG stream');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with a string expecting an PDF stream': function (test) {
    generateDiagram({
      input: '[nomnoml]is->[awesome]',
      format: 'pdf',
      resultType: 'stream'
    })
      .then(async function (result) {
        test.ok(true, 'succeeds');
        const buffer = await readStream(result);
        test.equal('%'.charCodeAt(0), buffer[0], 'returns "%" in a PDF stream');
        test.equal('P'.charCodeAt(0), buffer[1], 'returns "P" in a PDF stream');
        test.equal('D'.charCodeAt(0), buffer[2], 'returns "D" in a PDF stream');
        test.equal('F'.charCodeAt(0), buffer[3], 'returns "F" in a PDF stream');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with input and output streams': function (test) {
    const name = join(__dirname, 'piracy');
    generateDiagram({
      input: createReadStream(name + '_process-paths.nomnoml'),
      output: createWriteStream(name + '.png')
    })
      .then(function (/*result*/) {
        const output = statSync(name + '.png');
        test.ok(true, 'succeeds');
        test.ok(output.isFile() && output.size > 0, 'creates a PNG file');
        unlinkSync(name + '.png');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with input and output streams, with format=svg': function (test) {
    const name = join(__dirname, 'piracy');
    generateDiagram({
      input: createReadStream(name + '.nomnoml'),
      format: 'svg',
      output: createWriteStream(name + '.svg')
    })
      .then(function (/*result*/) {
        const output = statSync(name + '.svg');
        test.ok(true, 'succeeds');
        test.ok(output.isFile() && output.size > 0, 'creates a SVG file');
        unlinkSync(name + '.svg');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with correct input and output file names': function (test) {
    const name = join(__dirname, 'piracy');
    generateDiagram({
      inputFile: name + '_imports.nomnoml',
      output: name + '.png'
    })
      .then(function (/*result*/) {
        const output = statSync(name + '.png');
        test.ok(true, 'succeeds');
        test.ok(output.isFile() && output.size > 0, 'creates a PNG file');
        unlinkSync(name + '.png');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with correct input and output file names, with format=svg': function (test) {
    const name = join(__dirname, 'piracy');
    generateDiagram({
      inputFile: name + '.nomnoml',
      format: 'svg',
      output: name + '.svg'
    })
      .then(function (/*result*/) {
        const output = statSync(name + '.svg');
        test.ok(true, 'succeeds');
        test.ok(output.isFile() && output.size > 0, 'creates a SVG file');
        unlinkSync(name + '.svg');
        test.done();
      })
      .catch(function (error) {
        test.ok(false, error);
        test.done();
      });
  },

  'called with an invalid input file name': function (test) {
    generateDiagram({
      inputFile: 'nonsense'
    })
      .then(function () {
        test.ok(false, 'should fail');
        test.done();
      })
      .catch(function (error) {
        test.ok(true, 'fails');
        test.ok(error instanceof Error, 'returns an error');
        test.equal('ENOENT', error.code, 'describes the failure: "ENOENT"');
        test.done();
      });
  },

  'called with an invalid output file name': function (test) {
    const name = join(__dirname, 'piracy');
    generateDiagram({
      inputFile: name + '.nomnoml',
      output: ':/?*'
    })
      .then(function () {
        test.ok(false, 'should fail');
        test.done();
      })
      .catch(function (error) {
        test.ok(true, 'fails');
        test.ok(error instanceof Error, 'returns an error');
        test.equal('ENOENT', error.code, 'describes the failure: "ENOENT"');
        test.done();
      });
  }
};


test/problems_test.js
'use strict';

const { exec } = require('child_process');
const coverage = process.env.NOMNOML_CLI_COVERAGE;

require(coverage ? '../coverage/lib/generate' : '../lib/generate');

function checkFailure(test, error, stdout, stderr) {
  test.ok(error != null, 'fails');
  test.ok(stdout.length === 0, 'prints nothing to stdout');
  test.ok(stderr.length > 0, 'prints message to stderr');
  test.ok(error && error.code === 1, 'returns exit code 1');
  test.done();
}

exports.when = {
  'command-line tool called with wrong arguments': function (test) {
    exec('node ./bin/nomnoml -a',
        checkFailure.bind(null, test));
  },

  'command-line tool called with missing input file': function (test) {
    exec('node ./bin/nomnoml -i bad.nomnoml',
        checkFailure.bind(null, test));
  },

  'command-line tool called with malformed input file': function (test) {
    exec('node ./bin/nomnoml -i ./test/invalid.nomnoml',
        checkFailure.bind(null, test));
  }
};


Gruntfile.js
'use strict';

module.exports = function (grunt) {
  const { NOMNOML_CLI_COVERAGE: coverage } = process.env;

  require('time-grunt')(grunt);

  grunt.initConfig({
    eslint: {
      target:     [
        'Gruntfile.js',
        'lib/*.js',
        'bin/nomnoml',
        '<%= nodeunit.tests %>'
      ]
    },

    shell: {
      createPNG: {
        command: 'node ./bin/nomnoml -i ./test/piracy.nomnoml -o ./test/piracy.png'
      },
      createSVG: {
        command: 'node ./bin/nomnoml -i ./test/piracy.nomnoml -o ./test/piracy.svg -f svg'
      }
    },

    nodeunit: {
      tests:   ['test/*_test.js']
    },

    clean: {
      tests:    ['test/piracy.png'],
      coverage: ['coverage']
    },

    instrument: {
      files: 'lib/*.js',
      options: {
        lazy: true,
        basePath: 'coverage/'
      }
    },

    storeCoverage: {
      options: {
        dir: 'coverage'
      }
    },

    makeReport: {
      src: 'coverage/coverage.json',
      options: {
        type: 'lcov',
        dir: 'coverage',
        print: 'detail'
      }
    }
  });

  require('load-grunt-tasks')(grunt);

  grunt.registerTask('default', coverage ?
    ['eslint', 'clean', 'instrument', 'shell', 'nodeunit',
     'storeCoverage', 'makeReport'] :
    ['eslint', 'clean:tests', 'shell', 'nodeunit']);
};


lib/exporter.js
var Canvas = require('canvas'),
    fs = require('fs'),
    nomnoml = require('nomnoml'),
    Q = require('q'),
    stream = require('stream');

module.exports = function (options) {
  return Q.Promise(function (resolve, reject) {
    var inputText = '',
        inputStream;

    if (!options) {
      options = {};
    }
    if (typeof options === 'string') {
      options = {input: options};
    }

    if (options.input) {
      if (typeof input === 'string') {
        inputStream = new stream.Readable();
        inputStream.push(options.input);
      } else {
        inputStream = options.input;
      }
    } else if (options.inputFile) {
      inputStream = fs.createReadStream(options.inputFile);
    } else {
      throw new Error('Missing input');
    }

    inputStream.setEncoding('utf8');
    inputStream
      .on('data', function (chunk) {
        inputText += chunk;
      })
      .on('end', function () {
        var width = options.width || 640,
            height = options.height || 480,
            canvas = new Canvas(width, height),
            pngStream, outputStream;

        nomnoml.draw(canvas, inputText);

        if (options.output) {
          pngStream = canvas.pngStream();
          pngStream.on('error', function (error) {
            reject(error);
          });

          if (typeof options.output === 'string') {
            outputStream = fs.createWriteStream(options.output);
          } else {
            outputStream = options.output;
          }
          outputStream.on('error', function (error) {
            reject(error);
          });

          pngStream
            .pipe(outputStream)
            .on('end', function () {
              resolve();
            })
            .on('error', function (error) {
              reject(error);
            });
        } else {
          resolve(canvas.toBuffer());
        }
      })
      .on('error', function (error) {
        reject(error);
      });
  });
};

lib/generate.js
const { createCanvas } = require('canvas');
const { readFileSync, createWriteStream, createReadStream } = require('fs');
const { draw } = require('nomnoml');
const { isAbsolute, join, resolve, normalize, dirname } = require('path');
const { PassThrough } = require('stream');

const mimeTypes = {
  png: 'image/png',
  svg: 'image/svg+xml',
  jpg: 'image/jpeg',
  pdf: 'application/pdf'
};

// Expands file import directives recursively. A single file can be imported
// only once. It will be replaced by an empty string, if imported once more.
function preprocessInputText(inputText, currentDirectory) {
  const processDirectory = process.cwd();
  const importedFilePaths = [];

  // If the file name is an absolute path, returns it normailzed. If it is
  // a relative path starting with "./" or "../", appends it to the path
  // of the parent file, which the specified file is being expanded to.
  // Otherwise the relative path is  appended to the current directory.
  function completeFilePath(fileName, currentDirectory) {
    let directory;
    if (!isAbsolute(fileName)) {
      fileName = fileName.replace(/\\/g, '/');
      if (fileName.indexOf('./') === 0 ||
          fileName.indexOf('../') === 0) {
        directory = currentDirectory;
      } else {
        directory = processDirectory;
      }
      fileName = join(directory, fileName);
    }
    return resolve(normalize(fileName));
  }

  function expandFileImports(inputText, currentDirectory) {
    return inputText.replace(/#import: *(.*)/g, (all, fileName) => {
      fileName = completeFilePath(fileName, currentDirectory);
      if (importedFilePaths.indexOf(fileName) >= 0) {
        return '';
      }
      importedFilePaths.push(fileName);
      return expandFileImports(
          readFileSync(fileName, { encoding: 'utf8' }),
          dirname(fileName));
    });
  }

  return expandFileImports(inputText, currentDirectory || processDirectory);
}

module.exports = function (options) {
  return new Promise((resolve, reject) => {
    let inputText = '';
    let inputStream, input;

    function processInputText() {
      let {
        width = 640,
        height = 480,
        format = 'png',
        inputFile,
        output,
        resultType
      } = options;
      format = format.toLowerCase();

      const canvas = createCanvas(width, height,
        format === 'svg' ? 'svg' : format === 'pdf' ? 'pdf' : undefined);
      draw(canvas, preprocessInputText(
          inputText, inputFile && dirname(inputFile)));

      const buffer = canvas.toBuffer(mimeTypes[format]);
      if (output) {
        let outputStream;
        if (typeof output === 'string') {
          outputStream = createWriteStream(output);
        } else {
          outputStream = new PassThrough();
          outputStream.pipe(output);
        }
        outputStream
          .on('finish', () => resolve())
          .on('error', error => reject(error))
          .end(buffer);
      } else {
        if (resultType === 'stream') {
          const bufferStream = new PassThrough();
          bufferStream.end(buffer);
          resolve(bufferStream);
        }
        else {
          resolve(buffer);
        }
      }
    }

    if (!options) {
      throw new Error('Missing arguments');
    }
    if (typeof options === 'string') {
      options = { input: options };
    } else if (typeof options !== 'object') {
      throw new Error('Invalid arguments');
    }
    input = options.input;
    if (input) {
      if (typeof input === 'string') {
        inputText = input;
        return processInputText();
      }
      inputStream = input;
    } else if (options.inputFile) {
      inputStream = createReadStream(options.inputFile);
    } else {
      throw new Error('Missing input');
    }

    inputStream
      .on('data', chunk => inputText += chunk)
      .on('end', () => processInputText())
      .on('error', error => reject(error));
  });
};
